chapter-outline.md

```diff
@@ -0,0 +1,287 @@
+          that still give the student full freedom in how to implement the program
+          as long as the tests pass.
+          Maybe we can be creative and come up with something similar,
+          but it will be lower priority.
+        - Instead, we will have hints and solutions.
+          To encourage students to think for themselves
+          we could have a single button that generates hints on the first 1-2 clicks
+          and then shows the full solution in the next click.
+          I think this will make it a bit less tempting to just glance at the solution
+          instead of doing the exercise
+          and seeing part of the solution in the hint
+          might spark an idea before they click through the full solution
+          (a potential danger is that it is annoying for someone
+          who is quickly skimming solutions to review for a quiz,
+          so we will have to think about if that's an important use case
+          that we want to encourage/enable).
+        - One positive aspect of autograding that is harder to emulate
+          is that it gives some degree of personalized feedback to the student
+          based on the code they have so far.
+          A more "out there" idea in terms of achieving this without autograding,
+          could be one of the following:
+            - A peer feedback system where students get extra credit
+              for reviewing their peers code/answers in the textbook.
+              This would be difficult to track,
+              but could maybe create an additional social aspect
+              and it enables peer learning also for the reviewer.
+              However,
+              given the time available for MDS students,
+              this approach is likely unrealistic
+              and it would also not be helpful for students outside MDS
+              who want to go through the book.
+            - AI feedback where we provide an LLM
+              with both the student solution and the correct solution
+              and asks it to compare the two
+              and then help guide the student on the right path
+              similar to how an educator would do it
+              rather than giving the answer directly.
+              The LLM could ingest the content of the book,
+              with a higher fine-tuning weight on the current chapter/section maybe
+              (if a general one is not good enough)?
+              If this works well,
+              it would be fantastic
+              and students would essentially have access to a tutor right there in the textbook.
+              Both special made (Chat)GPTs and Khanmigo (an education bot from Khanacedmy)
+              could be promising candidates.
+              - I'm quite interested in this direction as it is a lower stakes environment
+                than labs/quizzes where students receive marks
+                and we could get useful feedback from students on how well it works.
```

Comment:
Maybe RAG could help here

Reply:
Maybe not yet possible when the student creates a chart (too low performance), but possible for text questions where we can feed it a textual solution? This could work well if we give a bad example of the chart and the student write down the text solution so that the LLM can use the solution to simulate e.g. a board outcome.

This is a bit experimental, but let's try to make prototype.

Maybe could use the VL json schema from Altair and have the model critique that instead of the visual itself? Would have to test performance to see where it is better.